// eas-cli/build/devices/manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountResolver = void 0;
const tslib_1 = require("tslib");
const assert_1 = tslib_1.__importDefault(require("assert"));
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const AppleTeamMutation_1 = require("../credentials/ios/api/graphql/mutations/AppleTeamMutation");
const AppleTeamQuery_1 = require("../credentials/ios/api/graphql/queries/AppleTeamQuery");
const log_1 = tslib_1.__importDefault(require("../log"));
const projectUtils_1 = require("../project/projectUtils");
const prompts_1 = require("../prompts");
const Account_1 = require("../user/Account");
const User_1 = require("../user/User");
const action_1 = tslib_1.__importDefault(require("./actions/create/action"));
const CREATE_COMMAND_DESCRIPTION = `This command lets you register your Apple devices (iPhones and iPads) for internal distribution of your app.
Internal distribution means that you won't need upload your app archive to App Store / Testflight.
Your app archive (.ipa) will be installable on your equipment as long as you sign your application with an adhoc provisiong profile.
The provisioning profile needs to contain the UDIDs (unique identifiers) of your iPhones and iPads.

First of all, please choose the Expo account under which you want to register your devices.
Later, authenticate with Apple and choose your desired Apple Team (if your Apple ID has access to multiple teams).`;
class DeviceManager {
    constructor(ctx) {
        this.ctx = ctx;
    }
    async createAsync() {
        log_1.default.log(chalk_1.default.green(CREATE_COMMAND_DESCRIPTION));
        log_1.default.addNewLineIfNone();
        const account = await this.resolveAccountAsync();
        const { team } = await this.ctx.appStore.ensureAuthenticatedAsync();
        const appleTeam = await ensureAppleTeamExistsAsync(account.id, {
            appleTeamIdentifier: team.id,
            appleTeamName: team.name,
        });
        const action = new action_1.default(account, appleTeam);
        await action.runAsync();
    }
    async resolveAccountAsync() {
        const resolver = new AccountResolver(this.ctx.exp, this.ctx.user);
        return await resolver.resolveAccountAsync();
    }
}
exports.default = DeviceManager;
class AccountResolver {
    constructor(exp, user) {
        this.exp = exp;
        this.user = user;
    }
    async resolveAccountAsync() {
        if (this.exp) {
            const account = await this.resolveProjectAccountAsync();
            if (account) {
                return account;
            }
        }
        return await this.promptForAccountAsync();
    }
    async resolveProjectAccountAsync() {
        (0, assert_1.default)(this.exp, 'expo config is not set');
        const projectAccountName = await (0, projectUtils_1.getProjectAccountNameAsync)(this.exp);
        const projectAccount = (0, Account_1.findAccountByName)(this.user.accounts, projectAccountName);
        if (!projectAccount) {
            log_1.default.warn(`Your account (${(0, User_1.getActorDisplayName)(this.user)}) doesn't have access to the ${chalk_1.default.bold(projectAccountName)} account`);
            return;
        }
        const useProjectAccount = await (0, prompts_1.confirmAsync)({
            message: `You're inside the project directory. Would you like to use ${chalk_1.default.underline(projectAccountName)} account?`,
        });
        return useProjectAccount ? projectAccount : undefined;
    }
    async promptForAccountAsync() {
        const choices = this.user.accounts.map(account => ({
            title: account.name,
            value: account,
        }));
        const { account } = await (0, prompts_1.promptAsync)({
            type: 'select',
            name: 'account',
            message: 'Which account to use?',
            choices,
        });
        return account;
    }
}
exports.AccountResolver = AccountResolver;
async function ensureAppleTeamExistsAsync(accountId, { appleTeamIdentifier, appleTeamName }) {
    const appleTeam = await AppleTeamQuery_1.AppleTeamQuery.getByAppleTeamIdentifierAsync(accountId, appleTeamIdentifier);
    if (appleTeam) {
        return appleTeam;
    }
    else {
        return await AppleTeamMutation_1.AppleTeamMutation.createAppleTeamAsync({
            appleTeamIdentifier,
            appleTeamName,
        }, accountId);
    }
}
