// eas-cli/build/credentials/ios/actions/BuildCredentialsUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAppLookupParamsFromContext = exports.getAppFromContext = exports.assignBuildCredentialsAsync = exports.getDistributionCertificateAsync = exports.getProvisioningProfileAsync = exports.getBuildCredentialsAsync = exports.getAllBuildCredentialsAsync = void 0;
const tslib_1 = require("tslib");
const nullthrows_1 = tslib_1.__importDefault(require("nullthrows"));
const projectUtils_1 = require("../../../project/projectUtils");
const Account_1 = require("../../../user/Account");
const AppleTeamUtils_1 = require("./AppleTeamUtils");
async function getAllBuildCredentialsAsync(ctx, app) {
    const appCredentials = await ctx.ios.getIosAppCredentialsWithBuildCredentialsAsync(app, {});
    if (!appCredentials) {
        return [];
    }
    return appCredentials.iosAppBuildCredentialsList;
}
exports.getAllBuildCredentialsAsync = getAllBuildCredentialsAsync;
async function getBuildCredentialsAsync(ctx, app, iosDistributionType) {
    const appCredentials = await ctx.ios.getIosAppCredentialsWithBuildCredentialsAsync(app, {
        iosDistributionType,
    });
    if (!appCredentials || appCredentials.iosAppBuildCredentialsList.length === 0) {
        return null;
    }
    const [buildCredentials] = appCredentials.iosAppBuildCredentialsList;
    return buildCredentials;
}
exports.getBuildCredentialsAsync = getBuildCredentialsAsync;
async function getProvisioningProfileAsync(ctx, app, iosDistributionType) {
    var _a;
    const buildCredentials = await getBuildCredentialsAsync(ctx, app, iosDistributionType);
    return (_a = buildCredentials === null || buildCredentials === void 0 ? void 0 : buildCredentials.provisioningProfile) !== null && _a !== void 0 ? _a : null;
}
exports.getProvisioningProfileAsync = getProvisioningProfileAsync;
async function getDistributionCertificateAsync(ctx, app, iosDistributionType) {
    var _a;
    const buildCredentials = await getBuildCredentialsAsync(ctx, app, iosDistributionType);
    return (_a = buildCredentials === null || buildCredentials === void 0 ? void 0 : buildCredentials.distributionCertificate) !== null && _a !== void 0 ? _a : null;
}
exports.getDistributionCertificateAsync = getDistributionCertificateAsync;
async function assignBuildCredentialsAsync(ctx, app, iosDistributionType, distCert, provisioningProfile, appleTeam) {
    const resolvedAppleTeam = (0, nullthrows_1.default)(appleTeam !== null && appleTeam !== void 0 ? appleTeam : (await (0, AppleTeamUtils_1.resolveAppleTeamIfAuthenticatedAsync)(ctx, app)));
    const appleAppIdentifier = await ctx.ios.createOrGetExistingAppleAppIdentifierAsync(app, resolvedAppleTeam);
    return await ctx.ios.createOrUpdateIosAppBuildCredentialsAsync(app, {
        appleTeam: resolvedAppleTeam,
        appleAppIdentifierId: appleAppIdentifier.id,
        appleDistributionCertificateId: distCert.id,
        appleProvisioningProfileId: provisioningProfile.id,
        iosDistributionType,
    });
}
exports.assignBuildCredentialsAsync = assignBuildCredentialsAsync;
function getAppFromContext(ctx) {
    ctx.ensureProjectContext();
    const projectName = ctx.exp.slug;
    const accountName = (0, projectUtils_1.getProjectAccountName)(ctx.exp, ctx.user);
    const account = (0, Account_1.findAccountByName)(ctx.user.accounts, accountName);
    if (!account) {
        throw new Error(`You do not have access to account: ${accountName}`);
    }
    return {
        account,
        projectName,
    };
}
exports.getAppFromContext = getAppFromContext;
function getAppLookupParamsFromContext(ctx, target) {
    const app = getAppFromContext(ctx);
    return { ...app, bundleIdentifier: target.bundleIdentifier };
}
exports.getAppLookupParamsFromContext = getAppLookupParamsFromContext;
