// eas-cli/build/credentials/ios/actions/DeviceUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatDeviceLabel = exports.chooseDevicesToDeleteAsync = exports.chooseDevicesAsync = void 0;
const AppleDevice_1 = require("../../../graphql/types/credentials/AppleDevice");
const prompts_1 = require("../.././../prompts");
async function chooseDevicesAsync(allDevices, preselectedDeviceIdentifiers = []) {
    const preselectedDeviceIdentifierSet = new Set(preselectedDeviceIdentifiers);
    const isSelected = (device) => preselectedDeviceIdentifierSet.size === 0 ||
        preselectedDeviceIdentifierSet.has(device.identifier);
    const { devices } = await (0, prompts_1.promptAsync)({
        type: 'multiselect',
        name: 'devices',
        message: 'Select devices for the adhoc build:',
        hint: '- Space to select. Return to submit',
        choices: allDevices.map(device => ({
            value: device,
            title: formatDeviceLabel(device),
            selected: isSelected(device),
        })),
        instructions: false,
        min: 1,
    });
    return devices;
}
exports.chooseDevicesAsync = chooseDevicesAsync;
async function chooseDevicesToDeleteAsync(allDevices) {
    const { devices } = await (0, prompts_1.promptAsync)({
        type: 'multiselect',
        name: 'devices',
        message: 'Which devices do you want to remove?',
        choices: allDevices.map(device => ({
            value: device,
            title: formatDeviceLabel(device),
            selected: false,
        })),
        instructions: false,
    });
    return devices;
}
exports.chooseDevicesToDeleteAsync = chooseDevicesToDeleteAsync;
function formatDeviceLabel(device) {
    var _a;
    const deviceDetails = formatDeviceDetails(device);
    return `${(_a = device.name) !== null && _a !== void 0 ? _a : device.identifier}${deviceDetails !== '' ? ` ${deviceDetails}` : ''}`;
}
exports.formatDeviceLabel = formatDeviceLabel;
function formatDeviceDetails(device) {
    let details = '';
    if (device.deviceClass) {
        details += AppleDevice_1.APPLE_DEVICE_CLASS_LABELS[device.deviceClass];
    }
    if (device.model) {
        details += details === '' ? device.model : ` ${device.model}`;
    }
    return details === '' ? details : `(${details})`;
}
