// @expo/eas-json/build/submit/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubmitProfileSchema = exports.IosSubmitProfileSchema = exports.AndroidSubmitProfileSchema = void 0;
const tslib_1 = require("tslib");
const joi_1 = tslib_1.__importDefault(require("joi"));
const types_1 = require("./types");
exports.AndroidSubmitProfileSchema = joi_1.default.object({
    serviceAccountKeyPath: joi_1.default.string(),
    track: joi_1.default.string()
        .valid(...Object.values(types_1.AndroidReleaseTrack))
        .default(types_1.AndroidReleaseTrack.internal),
    releaseStatus: joi_1.default.string()
        .valid(...Object.values(types_1.AndroidReleaseStatus))
        .default(types_1.AndroidReleaseStatus.completed),
    changesNotSentForReview: joi_1.default.boolean().default(false),
    applicationId: joi_1.default.string(),
});
exports.IosSubmitProfileSchema = joi_1.default.object({
    ascApiKeyPath: joi_1.default.string(),
    ascApiKeyId: joi_1.default.string(),
    ascApiKeyIssuerId: joi_1.default.string(),
    appleId: joi_1.default.string(),
    ascAppId: joi_1.default.string(),
    appleTeamId: joi_1.default.string(),
    sku: joi_1.default.string(),
    language: joi_1.default.string().default('en-US'),
    companyName: joi_1.default.string(),
    appName: joi_1.default.string(),
    bundleIdentifier: joi_1.default.string(),
});
exports.SubmitProfileSchema = joi_1.default.object({
    extends: joi_1.default.string(),
    android: exports.AndroidSubmitProfileSchema,
    ios: exports.IosSubmitProfileSchema,
});
