// eas-cli/build/analytics/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Analytics = exports.BuildEvent = exports.SubmissionEvent = void 0;
const rudderstackClient_1 = require("./rudderstackClient");
var SubmissionEvent;
(function (SubmissionEvent) {
    SubmissionEvent["SUBMIT_COMMAND"] = "submit cli submit command";
    SubmissionEvent["SUBMIT_COMMAND_ATTEMPT"] = "submit cli attempt";
    SubmissionEvent["SUBMIT_COMMAND_SUCCESS"] = "submit cli success";
    SubmissionEvent["SUBMIT_COMMAND_FAIL"] = "submit cli fail";
    SubmissionEvent["GATHER_CREDENTIALS_ATTEMPT"] = "submit cli gather credentials attempt";
    SubmissionEvent["GATHER_CREDENTIALS_SUCCESS"] = "submit cli gather credentials success";
    SubmissionEvent["GATHER_CREDENTIALS_FAIL"] = "submit cli gather credentials fail";
    SubmissionEvent["GATHER_ARCHIVE_ATTEMPT"] = "submit cli gather archive attempt";
    SubmissionEvent["GATHER_ARCHIVE_SUCCESS"] = "submit cli gather archive success";
    SubmissionEvent["GATHER_ARCHIVE_FAIL"] = "submit cli gather archive fail";
    SubmissionEvent["SUBMIT_REQUEST_ATTEMPT"] = "submit cli request attempt";
    SubmissionEvent["SUBMIT_REQUEST_SUCCESS"] = "submit cli request success";
    SubmissionEvent["SUBMIT_REQUEST_FAIL"] = "submit cli request fail";
})(SubmissionEvent = exports.SubmissionEvent || (exports.SubmissionEvent = {}));
var BuildEvent;
(function (BuildEvent) {
    BuildEvent["BUILD_COMMAND"] = "build cli build command";
    BuildEvent["PROJECT_UPLOAD_ATTEMPT"] = "build cli project upload attempt";
    BuildEvent["PROJECT_UPLOAD_SUCCESS"] = "build cli project upload success";
    BuildEvent["PROJECT_UPLOAD_FAIL"] = "build cli project upload fail";
    BuildEvent["GATHER_CREDENTIALS_ATTEMPT"] = "build cli gather credentials attempt";
    BuildEvent["GATHER_CREDENTIALS_SUCCESS"] = "build cli gather credentials success";
    BuildEvent["GATHER_CREDENTIALS_FAIL"] = "build cli gather credentials fail";
    BuildEvent["CONFIGURE_PROJECT_ATTEMPT"] = "build cli configure project attempt";
    BuildEvent["CONFIGURE_PROJECT_SUCCESS"] = "build cli configure project success";
    BuildEvent["CONFIGURE_PROJECT_FAIL"] = "build cli configure project fail";
    BuildEvent["BUILD_REQUEST_ATTEMPT"] = "build cli build request attempt";
    BuildEvent["BUILD_REQUEST_SUCCESS"] = "build cli build request success";
    BuildEvent["BUILD_REQUEST_FAIL"] = "build cli build request fail";
    BuildEvent["BUILD_STATUS_COMMAND"] = "build cli build status";
    BuildEvent["CREDENTIALS_SYNC_COMMAND"] = "build cli credentials sync command";
    BuildEvent["CREDENTIALS_SYNC_UPDATE_LOCAL_ATTEMPT"] = "build cli credentials sync update local attempt";
    BuildEvent["CREDENTIALS_SYNC_UPDATE_LOCAL_SUCCESS"] = "build cli credentials sync update local success";
    BuildEvent["CREDENTIALS_SYNC_UPDATE_LOCAL_FAIL"] = "build cli credentials sync update local fail";
    BuildEvent["CREDENTIALS_SYNC_UPDATE_REMOTE_ATTEMPT"] = "build cli credentials sync update remote attempt";
    BuildEvent["CREDENTIALS_SYNC_UPDATE_REMOTE_SUCCESS"] = "build cli credentials sync update remote success";
    BuildEvent["CREDENTIALS_SYNC_UPDATE_REMOTE_FAIL"] = "build cli credentials sync update remote fail";
    BuildEvent["ANDROID_KEYSTORE_CREATE"] = "build cli credentials keystore create";
})(BuildEvent = exports.BuildEvent || (exports.BuildEvent = {}));
class Analytics {
    static logEvent(name, properties) {
        (0, rudderstackClient_1.logEvent)(name, properties);
    }
}
exports.Analytics = Analytics;
