// eas-cli/build/credentials/manager/Actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IosActionType = exports.AndroidActionType = exports.Scope = void 0;
var Scope;
(function (Scope) {
    Scope[Scope["Project"] = 0] = "Project";
    Scope[Scope["Manager"] = 1] = "Manager";
    Scope[Scope["Account"] = 2] = "Account";
})(Scope = exports.Scope || (exports.Scope = {}));
var AndroidActionType;
(function (AndroidActionType) {
    AndroidActionType[AndroidActionType["ManageBuildCredentials"] = 0] = "ManageBuildCredentials";
    AndroidActionType[AndroidActionType["ManageFcm"] = 1] = "ManageFcm";
    AndroidActionType[AndroidActionType["ManageGoogleServiceAccountKey"] = 2] = "ManageGoogleServiceAccountKey";
    AndroidActionType[AndroidActionType["ManageCredentialsJson"] = 3] = "ManageCredentialsJson";
    AndroidActionType[AndroidActionType["GoBackToCaller"] = 4] = "GoBackToCaller";
    AndroidActionType[AndroidActionType["GoBackToHighLevelActions"] = 5] = "GoBackToHighLevelActions";
    AndroidActionType[AndroidActionType["CreateKeystore"] = 6] = "CreateKeystore";
    AndroidActionType[AndroidActionType["DownloadKeystore"] = 7] = "DownloadKeystore";
    AndroidActionType[AndroidActionType["RemoveKeystore"] = 8] = "RemoveKeystore";
    AndroidActionType[AndroidActionType["CreateFcm"] = 9] = "CreateFcm";
    AndroidActionType[AndroidActionType["RemoveFcm"] = 10] = "RemoveFcm";
    AndroidActionType[AndroidActionType["CreateGsaKey"] = 11] = "CreateGsaKey";
    AndroidActionType[AndroidActionType["UseExistingGsaKey"] = 12] = "UseExistingGsaKey";
    AndroidActionType[AndroidActionType["RemoveGsaKey"] = 13] = "RemoveGsaKey";
    AndroidActionType[AndroidActionType["SetUpGsaKey"] = 14] = "SetUpGsaKey";
    AndroidActionType[AndroidActionType["UpdateCredentialsJson"] = 15] = "UpdateCredentialsJson";
    AndroidActionType[AndroidActionType["SetUpBuildCredentialsFromCredentialsJson"] = 16] = "SetUpBuildCredentialsFromCredentialsJson";
})(AndroidActionType = exports.AndroidActionType || (exports.AndroidActionType = {}));
var IosActionType;
(function (IosActionType) {
    IosActionType[IosActionType["ManageCredentialsJson"] = 0] = "ManageCredentialsJson";
    IosActionType[IosActionType["ManageBuildCredentials"] = 1] = "ManageBuildCredentials";
    IosActionType[IosActionType["ManagePushKey"] = 2] = "ManagePushKey";
    IosActionType[IosActionType["ManageAscApiKey"] = 3] = "ManageAscApiKey";
    IosActionType[IosActionType["GoBackToCaller"] = 4] = "GoBackToCaller";
    IosActionType[IosActionType["GoBackToHighLevelActions"] = 5] = "GoBackToHighLevelActions";
    IosActionType[IosActionType["SetUpBuildCredentials"] = 6] = "SetUpBuildCredentials";
    IosActionType[IosActionType["SetUpBuildCredentialsFromCredentialsJson"] = 7] = "SetUpBuildCredentialsFromCredentialsJson";
    IosActionType[IosActionType["UpdateCredentialsJson"] = 8] = "UpdateCredentialsJson";
    IosActionType[IosActionType["UseExistingDistributionCertificate"] = 9] = "UseExistingDistributionCertificate";
    IosActionType[IosActionType["RemoveProvisioningProfile"] = 10] = "RemoveProvisioningProfile";
    IosActionType[IosActionType["CreateDistributionCertificate"] = 11] = "CreateDistributionCertificate";
    IosActionType[IosActionType["RemoveDistributionCertificate"] = 12] = "RemoveDistributionCertificate";
    IosActionType[IosActionType["SetUpPushKey"] = 13] = "SetUpPushKey";
    IosActionType[IosActionType["CreatePushKey"] = 14] = "CreatePushKey";
    IosActionType[IosActionType["UseExistingPushKey"] = 15] = "UseExistingPushKey";
    IosActionType[IosActionType["RemovePushKey"] = 16] = "RemovePushKey";
    IosActionType[IosActionType["SetUpAscApiKeyForSubmissions"] = 17] = "SetUpAscApiKeyForSubmissions";
    IosActionType[IosActionType["UseExistingAscApiKeyForSubmissions"] = 18] = "UseExistingAscApiKeyForSubmissions";
    IosActionType[IosActionType["CreateAscApiKeyForSubmissions"] = 19] = "CreateAscApiKeyForSubmissions";
    IosActionType[IosActionType["RemoveAscApiKey"] = 20] = "RemoveAscApiKey";
})(IosActionType = exports.IosActionType || (exports.IosActionType = {}));
