// eas-cli/build/credentials/manager/AndroidActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gsaKeyActions = exports.fcmActions = exports.buildCredentialsActions = exports.credentialsJsonActions = exports.highLevelActions = void 0;
const Actions_1 = require("./Actions");
exports.highLevelActions = [
    {
        value: Actions_1.AndroidActionType.ManageBuildCredentials,
        title: 'Keystore: Manage everything needed to build your project',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageFcm,
        title: 'Push Notifications: Manage your FCM API Key',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageGoogleServiceAccountKey,
        title: 'Google Service Account: Manage your Service Account Key',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.ManageCredentialsJson,
        title: 'credentials.json: Upload/Download credentials between EAS servers and your local json ',
        scope: Actions_1.Scope.Manager,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToCaller,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.credentialsJsonActions = [
    {
        value: Actions_1.AndroidActionType.UpdateCredentialsJson,
        title: 'Download credentials from EAS to credentials.json',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.SetUpBuildCredentialsFromCredentialsJson,
        title: 'Upload credentials from credentials.json to EAS',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.buildCredentialsActions = [
    {
        value: Actions_1.AndroidActionType.CreateKeystore,
        title: 'Set up a new keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.DownloadKeystore,
        title: 'Download existing keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.RemoveKeystore,
        title: 'Delete your keystore',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.fcmActions = [
    {
        value: Actions_1.AndroidActionType.CreateFcm,
        title: 'Upload an FCM API Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.RemoveFcm,
        title: 'Delete your FCM API Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
exports.gsaKeyActions = [
    {
        value: Actions_1.AndroidActionType.SetUpGsaKey,
        title: 'Set up a Google Service Account Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.CreateGsaKey,
        title: 'Upload a Google Service Account Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.UseExistingGsaKey,
        title: 'Use an existing Google Service Account Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.RemoveGsaKey,
        title: 'Delete a Google Service Account Key',
        scope: Actions_1.Scope.Project,
    },
    {
        value: Actions_1.AndroidActionType.GoBackToHighLevelActions,
        title: 'Go back',
        scope: Actions_1.Scope.Manager,
    },
];
