// eas-cli/build/build/ios/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformJob = void 0;
const tslib_1 = require("tslib");
const nullthrows_1 = tslib_1.__importDefault(require("nullthrows"));
const graphql_1 = require("../graphql");
function transformJob(job) {
    return {
        type: (0, graphql_1.transformWorkflow)(job.type),
        projectArchive: (0, graphql_1.transformProjectArchive)(job.projectArchive),
        projectRootDirectory: job.projectRootDirectory,
        releaseChannel: job.releaseChannel,
        updates: job.updates,
        secrets: transformIosSecrets(job.secrets),
        builderEnvironment: job.builderEnvironment,
        cache: job.cache,
        scheme: job.scheme,
        buildConfiguration: job.buildConfiguration,
        artifactPath: job.artifactPath,
        username: job.username,
        developmentClient: job.developmentClient,
        simulator: job.simulator,
        experimental: job.experimental,
    };
}
exports.transformJob = transformJob;
function transformIosSecrets(secrets) {
    var _a, _b, _c;
    const buildCredentials = [];
    for (const targetName of Object.keys((_a = secrets.buildCredentials) !== null && _a !== void 0 ? _a : {})) {
        buildCredentials.push({
            targetName,
            provisioningProfileBase64: (0, nullthrows_1.default)((_b = secrets.buildCredentials) === null || _b === void 0 ? void 0 : _b[targetName])
                .provisioningProfileBase64,
            distributionCertificate: (0, nullthrows_1.default)((_c = secrets.buildCredentials) === null || _c === void 0 ? void 0 : _c[targetName])
                .distributionCertificate,
        });
    }
    return {
        buildCredentials,
        environmentSecrets: secrets.env,
    };
}
