// eas-cli/build/credentials/ios/appstore/resolveCredentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deletePasswordAsync = exports.promptPasswordAsync = exports.resolveCredentialsAsync = void 0;
const tslib_1 = require("tslib");
const apple_utils_1 = require("@expo/apple-utils");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const fs = tslib_1.__importStar(require("fs-extra"));
const wrap_ansi_1 = tslib_1.__importDefault(require("wrap-ansi"));
const log_1 = tslib_1.__importStar(require("../../../log"));
const prompts_1 = require("../../../prompts");
const Keychain = tslib_1.__importStar(require("./keychain"));
/**
 * Get the username and possibly the password from the environment variables or the supplied options.
 * Password is optional because it's only needed for authentication, but not for re-authentication.
 *
 * @param options
 */
async function resolveCredentialsAsync(options) {
    const credentials = getAppleIdFromEnvironmentOrOptions(options);
    if (!credentials.username) {
        credentials.username = await promptUsernameAsync();
    }
    return credentials;
}
exports.resolveCredentialsAsync = resolveCredentialsAsync;
function getAppleIdFromEnvironmentOrOptions({ username, password, ...userCredentials }) {
    const passedAppleId = username || process.env.EXPO_APPLE_ID;
    // Only resolve the password if the username was provided.
    const passedAppleIdPassword = passedAppleId
        ? password || process.env.EXPO_APPLE_PASSWORD
        : undefined;
    return {
        ...userCredentials,
        username: passedAppleId,
        password: passedAppleIdPassword,
    };
}
async function promptUsernameAsync() {
    log_1.default.log('\u203A Log in to your Apple Developer account to continue');
    // Get the email address that was last used and set it as
    // the default value for quicker authentication.
    const lastAppleId = await getCachedUsernameAsync();
    const { username } = await (0, prompts_1.promptAsync)({
        type: 'text',
        name: 'username',
        message: `Apple ID:`,
        validate: (val) => val !== '',
        initial: lastAppleId !== null && lastAppleId !== void 0 ? lastAppleId : undefined,
    });
    if (username && username !== lastAppleId) {
        await cacheUsernameAsync(username);
    }
    return username;
}
async function cacheUsernameAsync(username) {
    // If a new email was used then store it as a suggestion for next time.
    // This functionality is disabled using the keychain mechanism.
    if (!Keychain.EXPO_NO_KEYCHAIN && username) {
        const cachedPath = apple_utils_1.JsonFileCache.usernameCachePath();
        await apple_utils_1.JsonFileCache.cacheAsync(cachedPath, { username });
    }
}
async function promptPasswordAsync({ username, }) {
    const cachedPassword = await getCachedPasswordAsync({ username });
    if (cachedPassword) {
        log_1.default.log(`\u203A Using password for ${username} from your local Keychain`);
        log_1.default.log(`  ${(0, log_1.learnMore)('https://docs.expo.dev/distribution/security#keychain')}`);
        return cachedPassword;
    }
    // https://docs.expo.dev/distribution/security/#apple-developer-account-credentials
    log_1.default.log((0, wrap_ansi_1.default)(chalk_1.default.bold(`\u203A The password is only used to authenticate with Apple and never stored on EAS servers`), process.stdout.columns || 80));
    log_1.default.log(`  ${(0, log_1.learnMore)('https://bit.ly/2VtGWhU')}`);
    const { password } = await (0, prompts_1.promptAsync)({
        type: 'password',
        name: 'password',
        message: () => `Password (for ${username}):`,
        validate: (val) => val !== '',
    });
    // TODO: Save only after the auth completes successfully.
    await cachePasswordAsync({ username, password });
    return password;
}
exports.promptPasswordAsync = promptPasswordAsync;
async function getCachedUsernameAsync() {
    var _a;
    if (Keychain.EXPO_NO_KEYCHAIN) {
        // Clear last used apple ID.
        await fs.remove(apple_utils_1.JsonFileCache.usernameCachePath());
        return null;
    }
    const cached = await apple_utils_1.JsonFileCache.getCacheAsync(apple_utils_1.JsonFileCache.usernameCachePath());
    const lastAppleId = (_a = cached === null || cached === void 0 ? void 0 : cached.username) !== null && _a !== void 0 ? _a : null;
    return typeof lastAppleId === 'string' ? lastAppleId : null;
}
/**
 * Returns the same prefix used by Fastlane in order to potentially share access between services.
 * [Cite. Fastlane](https://github.com/fastlane/fastlane/blob/f831062fa6f4b216b8ee38949adfe28fc11a0a8e/credentials_manager/lib/credentials_manager/account_manager.rb#L8).
 *
 * @param appleId email address
 */
function getKeychainServiceName(appleId) {
    return `deliver.${appleId}`;
}
async function deletePasswordAsync({ username, }) {
    const serviceName = getKeychainServiceName(username);
    const success = await Keychain.deletePasswordAsync({ username, serviceName });
    if (success) {
        log_1.default.log('\u203A Removed Apple ID password from the native Keychain');
    }
    return success;
}
exports.deletePasswordAsync = deletePasswordAsync;
async function getCachedPasswordAsync({ username, }) {
    // If the user opts out, delete the password.
    if (Keychain.EXPO_NO_KEYCHAIN) {
        await deletePasswordAsync({ username });
        return null;
    }
    const serviceName = getKeychainServiceName(username);
    return Keychain.getPasswordAsync({ username, serviceName });
}
async function cachePasswordAsync({ username, password }) {
    if (Keychain.EXPO_NO_KEYCHAIN) {
        log_1.default.log('\u203A Skip storing Apple ID password in the local Keychain.');
        return false;
    }
    log_1.default.log(`\u203A Saving Apple ID password to the local Keychain`);
    log_1.default.log(`  ${(0, log_1.learnMore)('https://docs.expo.dev/distribution/security#keychain')}`);
    const serviceName = getKeychainServiceName(username);
    return Keychain.setPasswordAsync({ username, password, serviceName });
}
