// eas-cli/build/build/android/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformJob = void 0;
const eas_build_job_1 = require("@expo/eas-build-job");
const generated_1 = require("../../graphql/generated");
const graphql_1 = require("../graphql");
function transformJob(job) {
    return {
        type: (0, graphql_1.transformWorkflow)(job.type),
        projectArchive: (0, graphql_1.transformProjectArchive)(job.projectArchive),
        projectRootDirectory: job.projectRootDirectory,
        releaseChannel: job.releaseChannel,
        updates: job.updates,
        secrets: job.secrets,
        builderEnvironment: job.builderEnvironment,
        cache: job.cache,
        gradleCommand: job.gradleCommand,
        artifactPath: job.artifactPath,
        username: job.username,
        buildType: job.buildType && transformBuildType(job.buildType),
        developmentClient: job.developmentClient,
        experimental: job.experimental,
    };
}
exports.transformJob = transformJob;
function transformBuildType(buildType) {
    if (buildType === eas_build_job_1.Android.BuildType.APK) {
        return generated_1.AndroidBuildType.Apk;
    }
    else {
        return generated_1.AndroidBuildType.AppBundle;
    }
}
