// eas-cli/build/build/android/prepareJob.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareJobAsync = void 0;
const tslib_1 = require("tslib");
const eas_build_job_1 = require("@expo/eas-build-job");
const path_1 = tslib_1.__importDefault(require("path"));
const slash_1 = tslib_1.__importDefault(require("slash"));
const projectUtils_1 = require("../../project/projectUtils");
const actions_1 = require("../../user/actions");
const vcs_1 = require("../../vcs");
const cacheDefaults = {
    disabled: false,
    customPaths: [],
    cacheDefaultPaths: true,
};
async function prepareJobAsync(ctx, jobData) {
    const username = (0, projectUtils_1.getUsername)(ctx.exp, await (0, actions_1.ensureLoggedInAsync)());
    const buildProfile = ctx.buildProfile;
    const projectRootDirectory = (0, slash_1.default)(path_1.default.relative(await (0, vcs_1.getVcsClient)().getRootPathAsync(), ctx.projectDir)) || '.';
    const { credentials } = jobData;
    const buildCredentials = credentials
        ? {
            buildCredentials: {
                keystore: {
                    dataBase64: credentials.keystore.keystore,
                    keystorePassword: credentials.keystore.keystorePassword,
                    keyAlias: credentials.keystore.keyAlias,
                    keyPassword: credentials.keystore.keyPassword,
                },
            },
        }
        : {};
    let buildType = ctx.buildProfile.buildType;
    if (!buildType && !buildProfile.gradleCommand && ctx.buildProfile.distribution === 'internal') {
        buildType = eas_build_job_1.Android.BuildType.APK;
    }
    const job = {
        type: ctx.workflow,
        platform: eas_build_job_1.Platform.ANDROID,
        projectRootDirectory,
        projectArchive: jobData.projectArchive,
        builderEnvironment: {
            image: buildProfile.image,
            node: buildProfile.node,
            yarn: buildProfile.yarn,
            ndk: buildProfile.ndk,
            expoCli: buildProfile.expoCli,
            env: buildProfile.env,
        },
        cache: {
            ...cacheDefaults,
            ...buildProfile.cache,
            clear: ctx.clearCache,
        },
        secrets: {
            ...buildCredentials,
        },
        releaseChannel: ctx.buildProfile.releaseChannel,
        updates: { channel: ctx.buildProfile.channel },
        developmentClient: buildProfile.developmentClient,
        gradleCommand: buildProfile.gradleCommand,
        artifactPath: buildProfile.artifactPath,
        buildType,
        username,
        experimental: {
            prebuildCommand: ctx.buildProfile.prebuildCommand,
        },
    };
    return (0, eas_build_job_1.sanitizeJob)(job);
}
exports.prepareJobAsync = prepareJobAsync;
