// eas-cli/build/credentials/android/AndroidCredentialsProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eas_build_job_1 = require("@expo/eas-build-job");
const eas_json_1 = require("@expo/eas-json");
const nullthrows_1 = tslib_1.__importDefault(require("nullthrows"));
const credentialsJsonReader = tslib_1.__importStar(require("../credentialsJson/read"));
const SetUpBuildCredentials_1 = require("./actions/SetUpBuildCredentials");
class AndroidCredentialsProvider {
    constructor(ctx, options) {
        this.ctx = ctx;
        this.options = options;
        this.platform = eas_build_job_1.Platform.ANDROID;
    }
    async getCredentialsAsync(src) {
        switch (src) {
            case eas_json_1.CredentialsSource.LOCAL:
                return await this.getLocalAsync();
            case eas_json_1.CredentialsSource.REMOTE:
                return await this.getRemoteAsync();
        }
    }
    async getRemoteAsync() {
        const setupBuildCredentialsAction = new SetUpBuildCredentials_1.SetUpBuildCredentials({ app: this.options.app });
        const buildCredentials = await setupBuildCredentialsAction.runAsync(this.ctx);
        return this.toAndroidCredentials(buildCredentials);
    }
    toAndroidCredentials(androidBuildCredentials) {
        var _a, _b, _c, _d, _e;
        return {
            keystore: {
                keystore: (0, nullthrows_1.default)((_a = androidBuildCredentials.androidKeystore) === null || _a === void 0 ? void 0 : _a.keystore),
                keystorePassword: (0, nullthrows_1.default)((_b = androidBuildCredentials.androidKeystore) === null || _b === void 0 ? void 0 : _b.keystorePassword),
                keyAlias: (0, nullthrows_1.default)((_c = androidBuildCredentials.androidKeystore) === null || _c === void 0 ? void 0 : _c.keyAlias),
                keyPassword: (_e = (_d = androidBuildCredentials.androidKeystore) === null || _d === void 0 ? void 0 : _d.keyPassword) !== null && _e !== void 0 ? _e : undefined,
            },
        };
    }
    async getLocalAsync() {
        return await credentialsJsonReader.readAndroidCredentialsAsync(this.ctx.projectDir);
    }
}
exports.default = AndroidCredentialsProvider;
