// eas-cli/build/build/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformIosEnterpriseProvisioning = exports.transformWorkflow = exports.transformMetadata = exports.transformProjectArchive = void 0;
const eas_build_job_1 = require("@expo/eas-build-job");
const generated_1 = require("../graphql/generated");
function transformProjectArchive(archiveSource) {
    if (archiveSource.type === eas_build_job_1.ArchiveSourceType.S3) {
        return {
            type: generated_1.ProjectArchiveSourceType.S3,
            bucketKey: archiveSource.bucketKey,
        };
    }
    else if (archiveSource.type === eas_build_job_1.ArchiveSourceType.URL) {
        return {
            type: generated_1.ProjectArchiveSourceType.Url,
            url: archiveSource.url,
        };
    }
    else {
        throw new Error(`Unsupported project archive source type: '${archiveSource.type}'`);
    }
}
exports.transformProjectArchive = transformProjectArchive;
function transformMetadata(metadata) {
    return {
        ...metadata,
        credentialsSource: metadata.credentialsSource && transformCredentialsSource(metadata.credentialsSource),
        distribution: metadata.distribution && transformDistribution(metadata.distribution),
        workflow: metadata.workflow && transformWorkflow(metadata.workflow),
        iosEnterpriseProvisioning: metadata.iosEnterpriseProvisioning &&
            transformIosEnterpriseProvisioning(metadata.iosEnterpriseProvisioning),
    };
}
exports.transformMetadata = transformMetadata;
function transformCredentialsSource(credentialsSource) {
    if (credentialsSource === 'local') {
        return generated_1.BuildCredentialsSource.Local;
    }
    else {
        return generated_1.BuildCredentialsSource.Remote;
    }
}
function transformDistribution(distribution) {
    if (distribution === 'internal') {
        return generated_1.DistributionType.Internal;
    }
    else if (distribution === 'simulator') {
        return generated_1.DistributionType.Simulator;
    }
    else {
        return generated_1.DistributionType.Store;
    }
}
function transformWorkflow(workflow) {
    if (workflow === eas_build_job_1.Workflow.GENERIC) {
        return generated_1.BuildWorkflow.Generic;
    }
    else {
        return generated_1.BuildWorkflow.Managed;
    }
}
exports.transformWorkflow = transformWorkflow;
function transformIosEnterpriseProvisioning(enterpriseProvisioning) {
    if (enterpriseProvisioning === 'adhoc') {
        return generated_1.BuildIosEnterpriseProvisioning.Adhoc;
    }
    else {
        return generated_1.BuildIosEnterpriseProvisioning.Universal;
    }
}
exports.transformIosEnterpriseProvisioning = transformIosEnterpriseProvisioning;
