// eas-cli/build/graphql/mutations/PublishMutation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublishMutation = void 0;
const tslib_1 = require("tslib");
const graphql_tag_1 = tslib_1.__importDefault(require("graphql-tag"));
const client_1 = require("../client");
exports.PublishMutation = {
    async getUploadURLsAsync(contentTypes) {
        const data = await (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .mutation((0, graphql_tag_1.default) `
            mutation GetSignedUploadMutation($contentTypes: [String!]!) {
              asset {
                getSignedAssetUploadSpecifications(assetContentTypes: $contentTypes) {
                  specifications
                }
              }
            }
          `, {
            contentTypes,
        })
            .toPromise());
        return data.asset.getSignedAssetUploadSpecifications;
    },
    async publishUpdateGroupAsync(publishUpdateGroupsInput) {
        const data = await (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .mutation((0, graphql_tag_1.default) `
            mutation UpdatePublishMutation($publishUpdateGroupsInput: [PublishUpdateGroupInput!]!) {
              updateBranch {
                publishUpdateGroups(publishUpdateGroupsInput: $publishUpdateGroupsInput) {
                  id
                  group
                  runtimeVersion
                  platform
                  manifestPermalink
                }
              }
            }
          `, { publishUpdateGroupsInput })
            .toPromise());
        return data.updateBranch.publishUpdateGroups;
    },
    async setCodeSigningInfoAsync(updateId, codeSigningInfo) {
        const data = await (0, client_1.withErrorHandlingAsync)(client_1.graphqlClient
            .mutation((0, graphql_tag_1.default) `
            mutation SetCodeSigningInfoMutation(
              $updateId: ID!
              $codeSigningInfo: CodeSigningInfoInput!
            ) {
              update {
                setCodeSigningInfo(updateId: $updateId, codeSigningInfo: $codeSigningInfo) {
                  id
                  group
                  awaitingCodeSigningInfo
                  codeSigningInfo {
                    keyid
                    alg
                    sig
                  }
                }
              }
            }
          `, { updateId, codeSigningInfo })
            .toPromise());
        return data.update.setCodeSigningInfo;
    },
};
