// @expo/eas-build-job/dist/android.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobSchema = exports.reactNativeVersionToDefaultBuilderImage = exports.builderBaseImages = exports.BuildType = void 0;
const joi_1 = __importDefault(require("joi"));
const common_1 = require("./common");
const KeystoreSchema = joi_1.default.object({
    dataBase64: joi_1.default.string().required(),
    keystorePassword: joi_1.default.string().allow('').required(),
    keyAlias: joi_1.default.string().required(),
    keyPassword: joi_1.default.string().allow(''),
});
var BuildType;
(function (BuildType) {
    BuildType["APK"] = "apk";
    BuildType["APP_BUNDLE"] = "app-bundle";
})(BuildType = exports.BuildType || (exports.BuildType = {}));
exports.builderBaseImages = [
    'default',
    'latest',
    'stable',
    'ubuntu-18.04-android-30-ndk-r19c',
    'ubuntu-20.04-android-30-ndk-r21e',
    'ubuntu-18.04-jdk-8-ndk-r19c',
    'ubuntu-18.04-jdk-11-ndk-r19c',
    'ubuntu-20.04-jdk-8-ndk-r21e',
    'ubuntu-20.04-jdk-11-ndk-r21e',
];
exports.reactNativeVersionToDefaultBuilderImage = {
    '>=0.68.0': 'ubuntu-18.04-jdk-11-ndk-r19c',
    '<0.68.0': 'ubuntu-18.04-jdk-8-ndk-r19c',
};
const BuilderEnvironmentSchema = joi_1.default.object({
    image: joi_1.default.string().valid(...exports.builderBaseImages),
    node: joi_1.default.string(),
    yarn: joi_1.default.string(),
    expoCli: joi_1.default.string(),
    ndk: joi_1.default.string(),
    env: common_1.EnvSchema,
});
exports.JobSchema = joi_1.default.object({
    type: joi_1.default.string()
        .valid(...Object.values(common_1.Workflow))
        .required(),
    projectArchive: common_1.ArchiveSourceSchema.required(),
    platform: joi_1.default.string().valid(common_1.Platform.ANDROID).required(),
    projectRootDirectory: joi_1.default.string().required(),
    releaseChannel: joi_1.default.string(),
    updates: joi_1.default.object({
        channel: joi_1.default.string(),
    }),
    secrets: joi_1.default.object({
        buildCredentials: joi_1.default.object({ keystore: KeystoreSchema.required() }),
        environmentSecrets: common_1.EnvSchema,
    }).required(),
    builderEnvironment: BuilderEnvironmentSchema,
    cache: common_1.CacheSchema.default(),
    developmentClient: joi_1.default.boolean(),
    gradleCommand: joi_1.default.string(),
    artifactPath: joi_1.default.string(),
    buildType: joi_1.default.string().valid(...Object.values(BuildType)),
    username: joi_1.default.string(),
    experimental: joi_1.default.object({
        prebuildCommand: joi_1.default.string(),
    }),
}).oxor('releaseChannel', 'updates.channel');
