// @expo/eas-build-job/dist/common.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheSchema = exports.EnvSchema = exports.ArchiveSourceSchema = exports.ArchiveSourceType = exports.Platform = exports.Workflow = void 0;
const joi_1 = __importDefault(require("joi"));
var Workflow;
(function (Workflow) {
    Workflow["GENERIC"] = "generic";
    Workflow["MANAGED"] = "managed";
})(Workflow = exports.Workflow || (exports.Workflow = {}));
var Platform;
(function (Platform) {
    Platform["ANDROID"] = "android";
    Platform["IOS"] = "ios";
})(Platform = exports.Platform || (exports.Platform = {}));
var ArchiveSourceType;
(function (ArchiveSourceType) {
    ArchiveSourceType["S3"] = "S3";
    ArchiveSourceType["URL"] = "URL";
    ArchiveSourceType["PATH"] = "PATH";
})(ArchiveSourceType = exports.ArchiveSourceType || (exports.ArchiveSourceType = {}));
exports.ArchiveSourceSchema = joi_1.default.object({
    type: joi_1.default.string()
        .valid(...Object.values(ArchiveSourceType))
        .required(),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.S3 }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.S3).required(),
        bucketKey: joi_1.default.string().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.URL }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.URL).required(),
        url: joi_1.default.string().uri().required(),
    }),
})
    .when(joi_1.default.object({ type: ArchiveSourceType.PATH }).unknown(), {
    then: joi_1.default.object({
        type: joi_1.default.string().valid(ArchiveSourceType.PATH).required(),
        path: joi_1.default.string().required(),
    }),
});
exports.EnvSchema = joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.string());
exports.CacheSchema = joi_1.default.object({
    disabled: joi_1.default.boolean().default(false),
    clear: joi_1.default.boolean().default(false),
    key: joi_1.default.string().allow('').max(128),
    cacheDefaultPaths: joi_1.default.boolean().default(true),
    customPaths: joi_1.default.array().items(joi_1.default.string()).default([]),
});
