// @expo/eas-build-job/dist/ios.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobSchema = exports.sdkVersionToDefaultBuilderImage = exports.builderBaseImages = void 0;
const joi_1 = __importDefault(require("joi"));
const common_1 = require("./common");
const TargetCredentialsSchema = joi_1.default.object().keys({
    provisioningProfileBase64: joi_1.default.string().required(),
    distributionCertificate: joi_1.default.object({
        dataBase64: joi_1.default.string().required(),
        password: joi_1.default.string().allow('').required(),
    }).required(),
});
const BuildCredentialsSchema = joi_1.default.object().pattern(joi_1.default.string().required(), TargetCredentialsSchema);
exports.builderBaseImages = [
    'default',
    'latest',
    'stable',
    'macos-big-sur-11.4-xcode-12.5',
    'macos-big-sur-11.4-xcode-13.0',
    'macos-monterey-12.1-xcode-13.2',
];
exports.sdkVersionToDefaultBuilderImage = {
    '<=42': 'macos-big-sur-11.4-xcode-12.5',
    '43': 'macos-big-sur-11.4-xcode-13.0',
};
const BuilderEnvironmentSchema = joi_1.default.object({
    image: joi_1.default.string().valid(...exports.builderBaseImages),
    node: joi_1.default.string(),
    yarn: joi_1.default.string(),
    expoCli: joi_1.default.string(),
    bundler: joi_1.default.string(),
    fastlane: joi_1.default.string(),
    cocoapods: joi_1.default.string(),
    env: common_1.EnvSchema,
});
exports.JobSchema = joi_1.default.object({
    type: joi_1.default.string()
        .valid(...Object.values(common_1.Workflow))
        .required(),
    projectArchive: common_1.ArchiveSourceSchema.required(),
    platform: joi_1.default.string().valid(common_1.Platform.IOS).required(),
    projectRootDirectory: joi_1.default.string().required(),
    releaseChannel: joi_1.default.string(),
    updates: joi_1.default.object({
        channel: joi_1.default.string(),
    }),
    secrets: joi_1.default.object({
        buildCredentials: BuildCredentialsSchema,
        environmentSecrets: common_1.EnvSchema,
    }).required(),
    builderEnvironment: BuilderEnvironmentSchema,
    cache: common_1.CacheSchema.default(),
    developmentClient: joi_1.default.boolean(),
    simulator: joi_1.default.boolean(),
    // generic
    scheme: joi_1.default.string(),
    buildConfiguration: joi_1.default.string(),
    artifactPath: joi_1.default.string(),
    // managed
    username: joi_1.default.string(),
    experimental: joi_1.default.object({
        prebuildCommand: joi_1.default.string(),
    }),
}).oxor('releaseChannel', 'updates.channel');
