// @expo/eas-build-job/dist/job.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanitizeJob = exports.JobSchema = void 0;
const joi_1 = __importDefault(require("joi"));
const semver_1 = __importDefault(require("semver"));
const common_1 = require("./common");
const Android = __importStar(require("./android"));
const Ios = __importStar(require("./ios"));
exports.JobSchema = joi_1.default.object({
    platform: joi_1.default.string()
        .valid(...Object.values(common_1.Platform))
        .required(),
})
    .when(joi_1.default.object({ platform: common_1.Platform.ANDROID }).unknown(), { then: Android.JobSchema })
    .when(joi_1.default.object({ platform: common_1.Platform.IOS }).unknown(), { then: Ios.JobSchema });
function sanitizeJob(rawJob, { sdkVersion, reactNativeVersion } = {}) {
    const { value, error } = exports.JobSchema.validate(rawJob, {
        stripUnknown: true,
        convert: true,
        abortEarly: false,
    });
    const job = value;
    if (job.platform === common_1.Platform.IOS) {
        setIosBuilderImageForManagedJob(job, sdkVersion);
    }
    else if (job.platform === common_1.Platform.ANDROID) {
        setAndroidBuilderImage(job, reactNativeVersion);
    }
    if (error) {
        throw error;
    }
    else {
        return job;
    }
}
exports.sanitizeJob = sanitizeJob;
function setAndroidBuilderImage(job, reactNativeVersion) {
    var _a;
    if (((_a = job.builderEnvironment) === null || _a === void 0 ? void 0 : _a.image) || !reactNativeVersion) {
        return;
    }
    const ranges = Object.keys(Android.reactNativeVersionToDefaultBuilderImage);
    const matchingRange = ranges.find((range) => semver_1.default.satisfies(reactNativeVersion, range));
    if (!matchingRange) {
        return;
    }
    const image = Android.reactNativeVersionToDefaultBuilderImage[matchingRange];
    job.builderEnvironment = Object.assign({ image }, job.builderEnvironment);
}
function setIosBuilderImageForManagedJob(job, sdkVersion) {
    var _a;
    if (job.type !== common_1.Workflow.MANAGED || ((_a = job.builderEnvironment) === null || _a === void 0 ? void 0 : _a.image) || !sdkVersion) {
        return;
    }
    const ranges = Object.keys(Ios.sdkVersionToDefaultBuilderImage);
    const matchingRange = ranges.find((range) => semver_1.default.satisfies(sdkVersion, range));
    if (!matchingRange) {
        return;
    }
    const image = Ios.sdkVersionToDefaultBuilderImage[matchingRange];
    job.builderEnvironment = Object.assign({ image }, job.builderEnvironment);
}
