// @expo/eas-build-job/dist/logs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogMarker = exports.BuildPhase = void 0;
var BuildPhase;
(function (BuildPhase) {
    BuildPhase["UNKNOWN"] = "UNKNOWN";
    BuildPhase["QUEUE"] = "QUEUE";
    BuildPhase["SPIN_UP_BUILDER"] = "SPIN_UP_BUILDER";
    BuildPhase["BUILDER_INFO"] = "BUILDER_INFO";
    BuildPhase["READ_APP_CONFIG"] = "READ_APP_CONFIG";
    BuildPhase["READ_PACKAGE_JSON"] = "READ_PACKAGE_JSON";
    BuildPhase["RUN_EXPO_DOCTOR"] = "RUN_EXPO_DOCTOR";
    BuildPhase["SET_UP_BUILD_ENVIRONMENT"] = "SET_UP_BUILD_ENVIRONMENT";
    BuildPhase["START_BUILD"] = "START_BUILD";
    BuildPhase["INSTALL_CUSTOM_TOOLS"] = "INSTALL_CUSTOM_TOOLS";
    BuildPhase["PREPARE_PROJECT"] = "PREPARE_PROJECT";
    BuildPhase["RESTORE_CACHE"] = "RESTORE_CACHE";
    BuildPhase["INSTALL_DEPENDENCIES"] = "INSTALL_DEPENDENCIES";
    BuildPhase["PREBUILD"] = "PREBUILD";
    BuildPhase["PREPARE_CREDENTIALS"] = "PREPARE_CREDENTIALS";
    BuildPhase["CONFIGURE_EXPO_UPDATES"] = "CONFIGURE_EXPO_UPDATES";
    BuildPhase["SAVE_CACHE"] = "SAVE_CACHE";
    BuildPhase["UPLOAD_ARTIFACTS"] = "UPLOAD_ARTIFACTS";
    BuildPhase["PREPARE_ARTIFACTS"] = "PREPARE_ARTIFACTS";
    BuildPhase["CLEAN_UP_CREDENTIALS"] = "CLEAN_UP_CREDENTIALS";
    BuildPhase["COMPLETE_BUILD"] = "COMPLETE_BUILD";
    BuildPhase["FAIL_BUILD"] = "FAIL_BUILD";
    // ANDROID
    BuildPhase["FIX_GRADLEW"] = "FIX_GRADLEW";
    BuildPhase["RUN_GRADLEW"] = "RUN_GRADLEW";
    // IOS
    BuildPhase["INSTALL_PODS"] = "INSTALL_PODS";
    BuildPhase["CONFIGURE_XCODE_PROJECT"] = "CONFIGURE_XCODE_PROJECT";
    BuildPhase["RUN_FASTLANE"] = "RUN_FASTLANE";
    // HOOKS
    BuildPhase["PRE_INSTALL_HOOK"] = "PRE_INSTALL_HOOK";
    BuildPhase["POST_INSTALL_HOOK"] = "POST_INSTALL_HOOK";
    BuildPhase["PRE_UPLOAD_ARTIFACTS_HOOK"] = "PRE_UPLOAD_ARTIFACTS_HOOK";
})(BuildPhase = exports.BuildPhase || (exports.BuildPhase = {}));
var LogMarker;
(function (LogMarker) {
    LogMarker["START_PHASE"] = "START_PHASE";
    LogMarker["END_PHASE"] = "END_PHASE";
})(LogMarker = exports.LogMarker || (exports.LogMarker = {}));
